File: src/main/java/ru/rumal/wishlist/exception/ApiError.java
package ru.rumal.wishlist.exception;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;
import org.springframework.http.HttpStatus;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.List;

@Data
public class ApiError {

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd-MM-yyyy hh:mm:ss")
    private final LocalDateTime timestamp;
    private HttpStatus status;
    private String message;
    private List<ApiValidationError> subErrors;

    private ApiError() {
        this.timestamp = LocalDateTime.now();
    }

    public ApiError(HttpStatus status,
                    String message) {
        this();
        this.status = status;
        this.message = message;
    }

    public static void writeApiError(HttpServletResponse response,
                                     ObjectMapper mapper,
                                     ApiError apiError) throws IOException {
        mapper.writeValue(response.getOutputStream(), apiError);
    }
}

File: src/main/resources/js/model/ApiError.js
class ApiError {

    constructor(options) {
        this.timestamp = options.timestamp ? options.timestamp : null;
        this.message = options.message ? options.message : null;
        this.status = options.status ? options.status : null;
        this.subErrors = [];
        if (options.subErrors) {
            for (const item of options.subErrors) {
                this.subErrors.push(new SubError(item))
            }
        }
    }
}

class SubError {
    constructor(options) {
        this.object = options.object ? options.object : null;
        this.field = options.field ? options.field : null;
        this.rejectedValue = options.rejectedValue ? options.rejectedValue : null;
        this.message = options.message ? options.message : null;
    }
}

export {ApiError, SubError}